cli/command/service/ps.go

```diff
@@ -74,29 +75,43 @@ func runPS(dockerCli *command.DockerCli, opts psOptions) error {
 		return err
 	}
 
+	var errs []string
+	serviceCount := 0
+loop:
+	// Match services by 1. Full ID, 2. Full name, 3. ID prefix. An error is returned if the ID-prefix match is ambiguous
 	for _, service := range opts.services {
-		serviceCount := 0
-		// Lookup by ID/Prefix
-		for _, serviceEntry := range serviceByIDList {
-			if strings.HasPrefix(serviceEntry.ID, service) {
-				filter.Add("service", serviceEntry.ID)
+		for _, s := range serviceByIDList {
+			if s.ID == service {
+				filter.Add("service", s.ID)
 				serviceCount++
+				continue loop
 			}
 		}
-
-		// Lookup by Name/Prefix
-		for _, serviceEntry := range serviceByNameList {
-			if strings.HasPrefix(serviceEntry.Spec.Annotations.Name, service) {
-				filter.Add("service", serviceEntry.ID)
+		for _, s := range serviceByNameList {
+			if s.Spec.Annotations.Name == service {
+				filter.Add("service", s.ID)
 				serviceCount++
+				continue loop
 			}
 		}
-		// If nothing has been found, return immediately.
-		if serviceCount == 0 {
-			return errors.Errorf("no such services: %s", service)
+		matches := 0
+		for _, s := range serviceByIDList {
+			if strings.HasPrefix(s.ID, service) {
+				filter.Add("service", s.ID)
+				serviceCount++
+				matches++
+				if matches > 1 {
+					return errors.Errorf("multiple services found with provided prefix: %s", service)
+				}
+			}
+		}
+		if matches == 0 {
+			errs = append(errs, fmt.Sprintf("no such service: %s", service))
```

Comment:
Feel free to ignore this, but I try to avoid using `Sprintf` for simple string concatenation. With the `+` operator, there isn't a chance of proving the wrong number or type of arguments to `Sprintf`.